cornac/eval_methods/propensity_stratified_evaluation.py

```diff
@@ -0,0 +1,514 @@
+        self.train_size, self.val_size, self.test_size = RatioSplit.validate_size(
+            val_size, test_size, len(self._data))
+        self._split()
+
+    def _eval(self, model, test_set, val_set, user_based, props=None, self_normalized=True):
+
+        metric_avg_results = OrderedDict()
+        metric_user_results = OrderedDict()
+
+        avg_results, user_results = rating_eval(
+            model=model,
+            metrics=self.rating_metrics,
+            test_set=test_set,
+            user_based=user_based,
+        )
+        for i, mt in enumerate(self.rating_metrics):
+            metric_avg_results[mt.name] = avg_results[i]
+            metric_user_results[mt.name] = user_results[i]
+
+        avg_results, user_results = ranking_eval(
+            model=model,
+            metrics=self.ranking_metrics,
+            train_set=self.train_set,
+            test_set=test_set,
+            val_set=val_set,
+            rating_threshold=self.rating_threshold,
+            exclude_unknowns=self.exclude_unknowns,
+            verbose=self.verbose,
+            props=props,
+            self_normalized=self_normalized
+        )
+        for i, mt in enumerate(self.ranking_metrics):
+            metric_avg_results[mt.name] = avg_results[i]
+            metric_user_results[mt.name] = user_results[i]
+
+        return Result(model.name, metric_avg_results, metric_user_results)
+
+    def _split(self):
+        data_idx = self.rng.permutation(len(self._data))
+        train_idx = data_idx[:self.train_size]
+        test_idx = data_idx[-self.test_size:]
+        val_idx = data_idx[self.train_size:-self.test_size]
+
+        train_data = safe_indexing(self._data, train_idx)
+        test_data = safe_indexing(self._data, test_idx)
+        val_data = safe_indexing(self._data, val_idx) if len(
+            val_idx) > 0 else None
+
+        self._build_stratified_datasets(train_data=train_data,
+                                        test_data=test_data,
+                                        val_data=val_data)
+
+    def _estimate_propensities(self):
+
+        # find the item's frequencies
+        item_freq = defaultdict(int)
+        for u, i, r in self._data:
+            item_freq[i] += 1
+
+        # fit the exponential param
+        data = np.array([e for e in item_freq.values()], dtype=np.float)
+        results = powerlaw.Fit(data, discrete=True,
+                               fit_method='Likelihood')
+        alpha = results.power_law.alpha
+        fmin = results.power_law.xmin
+
+        if self.verbose:
+            print('Powerlaw exponential estimates: %f, min=%d' % (alpha, fmin))
+
+        # replace raw frequencies with the estimated propensities
+        for k, v in item_freq.items():
+            if v > fmin:
+                item_freq[k] = pow(v, alpha)
+
+        return item_freq  # user-independent propensity estimations
+
+    def _build_stratified_datasets(self, train_data, test_data, val_data):
```

Comment:
The first part of this function is identical to the [`_build_datasets()` function in the `BaseMethod`](https://github.com/PreferredAI/cornac/blob/4757ed1f1cacbd43548bfa6cd67855625987796d/cornac/eval_methods/base_method.py#L461). Can it be reused?

I can think of writing a function just to create `stratified_datasets` from `test_data`, and call it after the `_build_datasets()`. That would be more synchronized with the current codebase.